### ISSUE TITLE
SimpleMessageGroup constructor has unnecessary protected access

### ISSUE THREAD

**Issue by asarkar** [spring-projects/spring-integration#2301]
```
protected SimpleMessageGroup(Collection<Message<?>> internalStore, Collection<? extends Message<?>> messages,
			Object groupId, long timestamp, boolean complete, boolean storePreLoaded)
```

This makes it very difficult to implement a custom `MessageGroupFactory` that instantiates a `SimpleMessageGroup` with a different type of `Collection`. The workaround is to put the `MyMessageGroupFactory` in the same package, but that is a nasty hack.

**Comment by artembilan** [comment]
The idea behind that `protected` ctor that together with the custom `MessageGroupFactory` you would instantiate your own `MessageGroup` implementation. Even if it is just simple extension of the `SimpleMessageGroup`. 

Just to avoid confuse at runtime and be able to determine by the target type of the `MessageGroup` its source factory.

**Comment by asarkar** [comment]
@artembilan Why'd I be forced to create my own `MessageGroup` if `SimpleMessageGroup` works for me? What is the design though behind that?

**Comment by artembilan** [comment]
`SimpleMessageGroupFactory` -> `SimpleMessageGroup`.
`MyMessageGroupFactory` -> `MyMessageGroup`.

The `MyMessageGroup` might be just like `extends SimpleMessageGroup` and you can make that ctor as `public`. That's just a canonical factory pattern. 

Would be great though to hear why would one need the custom `MessageGroupFactory` per se? 
Why just don't improve an existing `SimpleMessageGroupFactory` via PullRequest if we are missing anything there?

Thanks

**Comment by asarkar** [comment]
> why would one need the custom `MessageGroupFactory`

Because of this part of the doc

[Aggregating Large Groups](https://docs.spring.io/spring-integration/docs/5.0.0.RELEASE/reference/html/messaging-routing-chapter.html#_aggregating_large_groups)

**Comment by artembilan** [comment]
Well, but there is some answer to your problem:

> a SimpleMessageGroupFactory is provided so you can select the Collection that best suits your needs. 

And at the same time you don't answer to my question regarding this part:

> You can also provide your own factory implementation to create some other Collection<Message<?>>.

In time writing that solution we had only particular requirements. Right now you come to us with some problem, which might be really fixed on the Framework level for reuse in the community. But we still need to know what drives you to go the way with your own `MessageGroupFactory`.  What is the problem to answer to this question? Thanks

**Comment by asarkar** [comment]
Sorry if I didn't understand your question earlier, let me try again. I'm aggregating a `Flux` of about 12 million records , and want to make sure that I get the last drop of performance out of the code. My message payload is `String`, and I don't need deduplication logic, although I want partial release. Given that, I think a `List` type would be more appropriate than a `HashSet` because there is no need for a message to have unique identity in the group. However, there is no `List` implementation available out of the box, so my approach was to create my own.

I'm open for other ideas if the above is not the optimal approach.

**Comment by artembilan** [comment]
OK! Now it's clear. Thank you!

So, Let's consider that as an improvement in the Framework!

Well, I even may reconsider that `protected` ctor as well. Just to save some time in the future avoiding similar discussion. 

I think we even might think about something like `Flux<Message<?>> MessageGroup.getMessagesFlux()`, but that's already a different story...

**Comment by garyrussell** [comment]
>want to get the last drop of performance 

> although I want partial release.

You might find that a `LinkedList` will perform better than an `ArrayList` since each removal will require a shift.

**Comment by asarkar** [comment]
After some testing, I didn't find significant difference between the  out-of-the-box implementation and my custom implementation, so I dropped my custom code. But point noted, when I do performance fine-tuning, I may need to use a `LinkedList`
